packages/context/src/handler-chain.ts

```diff
@@ -0,0 +1,210 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {BindingFilter} from './binding-filter';
+import {BindingAddress} from './binding-key';
+import {BindingComparator} from './binding-sorter';
+import {Context} from './context';
+import {transformValueOrPromise, ValueOrPromise} from './value-promise';
+
+const debug = debugFactory('loopback:context:handler-chain');
+
+/**
+ * Return value for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationResult = any;
+
+/**
+ * Array of arguments for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationArgs = any[];
+
+/**
+ * `next` function
+ */
+export type Next = () => ValueOrPromise<InvocationResult>;
+
+/**
+ * A handler function to be invoked in a chain for the given context
+ *
+ * @param context - Context object
+ * @param next - A function to proceed with downstream handlers or the target
+ *
+ * @returns The invocation result as a value or promise
+ */
+export type Handler<C extends Context = Context> = (
```

Comment:
> I'm concerned that our users will mistakenly assume it's Express/Koa middleware.
> `Handler` is used in https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern.

Well, the interface does describe Koa middleware too, doesn't it? I feel it's actually a good thing if people assume this type is for Koa-like Middleware, because it will make it easier for them to understand the concept.

Would `GenericMiddleware<T>` alleviate your concern?

Reply:
Can we find any other name, one that has no connotations to "route handlers" and "express middleware"?